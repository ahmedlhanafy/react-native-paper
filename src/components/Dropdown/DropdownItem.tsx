import React, {useContext} from "react";
import {MenuItemProps} from "react-native-paper";
import MenuItem from "../Menu/MenuItem";
import DropdownContext from "./DropdownContext";

export interface Props extends Omit<MenuItemProps, 'onPress'> {
  value: string;
}

const DropdownItem = (props: Props) => {
  const dropdownContext = useContext(DropdownContext);

  return <MenuItem style={{maxWidth: undefined}} onPress={() => dropdownContext.onChange?.(props.value)} {...props}/>;
}


export default DropdownItem;
