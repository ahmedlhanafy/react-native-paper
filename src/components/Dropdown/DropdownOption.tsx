import * as React from 'react';
import type { StyleProp, ViewStyle } from 'react-native';
import { withTheme } from '../../core/theming';

export interface Props<T> {
  /**
   * @optional
   */
  style?: StyleProp<ViewStyle>;
  /**
   * @optional
   */
  theme?: ReactNativePaper.Theme;
  /**
   * Value represented by the dropdown option
   */
  value: T | null;
  /**
   * Required unique key
   */
  optionKey: React.Key;
  /**
   * Title that will be shown in the dropdown menu when opened
   */
  title?: React.ReactNode;
  /**
   * Description will appear below the title when the menu is open
   */
  description?: React.ReactNode;
  /**
   * Label that will show if the current option is selected.
   * If undefined the title prop will be used instead
   */
  label?: string;
  /**
   * Callback which returns a React element to display on the right side.
   */
  right?: (props: {
    color: string;
    style?: {
      marginRight: number;
      marginVertical?: number;
    };
  }) => React.ReactNode;
  /**
   * Callback which returns a React element to display on the left side.
   */
  left?: (props: {
    color: string;
    style: {
      marginLeft: number;
      marginRight: number;
      marginVertical?: number;
    };
  }) => React.ReactNode;
}

/**
 * A dropdown option can only be rendered within a Dropdown component
 *
 * ## Usage
 * ```js
 * import * as React from 'react';
 * import {
 *   Dropdown,
 *   IconButton,
 *   Provider,
 *   Text,
 *   Title,
 * } from 'react-native-paper';
 *
 * const MyComponent = () => {
 *   return (
 *     <Provider>
 *       <Dropdown>
 *         <Dropdown.Option
 *          value={1}
 *          optionKey={1}
 *          title="Option 1"
 *          left={() => <IconButton icon="emoticon"/>}
 *          label="Option 1 is selected!"
 *         />
 *       </Dropdown>
 *     </Provider>
 *   );
 * };
 *
 * export default MyComponent;
 * ```
 */
class DropdownOption<T> extends React.Component<
  Props<T> & { theme: ReactNativePaper.Theme }
> {
  static displayName = 'Dropdown.Option';

  // eslint-disable-next-line react/require-render-return
  render(): React.ReactNode {
    throw new Error('<DropdownOption/> must be a direct child to <Dropdown/>');
  }
}

export default withTheme(DropdownOption);

// @component-docs ignore-next-line
export { DropdownOption };
