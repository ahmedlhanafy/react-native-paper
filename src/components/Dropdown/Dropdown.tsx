import * as React from 'react';
import {
  Platform,
  ScrollView,
  StyleSheet,
  TouchableWithoutFeedback,
  View,
} from 'react-native';
import * as List from '../List/List';
import Surface from '../Surface';
import TouchableRipple from '../TouchableRipple/TouchableRipple';
import IconButton from '../IconButton';
import Portal from '../Portal/Portal';
import Dialog from '../Dialog/Dialog';
import Text from '../Typography/Text';
import { withTheme } from '../../core/theming';
import DropdownOption, { Props as OptionProps } from './DropdownOption';

type Props<T> = {
  /**
   * Placeholder label when there is no selected value
   */
  placeholder?: string;
  /**
   * Required flag removes the possibility for the user to unselect a value once it's selected
   */
  required?: boolean;
  /**
   * One or multiple Dropdown.Option elements
   */
  children?:
    | React.ReactElement<OptionProps<T>>[]
    | React.ReactElement<OptionProps<T>>;
  /**
   * Only for web, maximum height of the dropdown popup
   */
  maxHeight?: number;
  /**
   * Triggered when on of the options is selected or when unselected
   * @param selected the selected option value or null if nothing is selected
   */
  onSelect?: (selected: T | null) => void;
  /**
   * The option having this key will be selected
   * If undefined the currently selected option will remain selected
   */
  selectedKey?: React.Key | null;
  /**
   * This message will appear when the required prop is set to true and the number of children is zero
   */
  emptyDropdownMessage?: string;
  /**
   * The label of the unselect option
   */
  unselectLabel?: string;
  /**
   * Callback which returns a React element to display on the left side of the unselect option
   */
  unselectIcon?: (props: {
    color: string;
    style: {
      marginLeft: number;
      marginRight: number;
      marginVertical?: number;
    };
  }) => React.ReactNode;
  /**
   * @optional
   */
  theme: ReactNativePaper.Theme;
};

type State = {
  coordinates: { top: number; left: number; width: number };
  isOpen: boolean;
  selectedKey: React.Key | null;
};

const DEFAULT_EMPTY_DROPDOWN_LABEL = 'No available options';
const DEFAULT_PLACEHOLDER_LABEL = 'Select an option';
const DEFAULT_MAX_HEIGHT = 350;
const DROPDOWN_NULL_OPTION_KEY = 'DROPDOWN_NULL_OPTION_KEY';

const DropdownChild = <T extends any>(props: {
  props: OptionProps<T>;
  onPress: () => void;
}) => (
  <List.Item
    style={props.props.style}
    title={props.props.title}
    description={props.props.description}
    right={props.props.right}
    onPress={props.onPress}
    left={props.props.left}
    theme={props.props.theme}
  />
);

/**
 * Dropdowns present a list of options from which a user can select one option.
 * A selected option can represent a value in a form, or can be used as an action to filter or sort existing content.
 *
 * <div class="screenshots">
 *   <figure>
 *     <img class="medium" src="screenshots/dropdown.png" />
 *   </figure>
 * </div>
 *
 * ## Usage
 * ```js
 * import * as React from 'react';
 * import { Dropdown, Provider, Text, Title } from 'react-native-paper';
 *
 * const MyComponent = () => {
 *   const [selectedOption, setSelectedOption] = React.useState(null);
 *   const options = [
 *     {id: 1, name: 'Cookie', calories: 502},
 *     {id: 2, name: 'Candy', calories: 535},
 *   ];
 *
 *   return (
 *     <Provider>
 *       <Dropdown onSelect={setSelectedOption}>
 *         {options.map(option => (
 *           <Dropdown.Option
 *            value={option}
 *            optionKey={value.id}
 *            title={value.name}
 *           />
 *         ))}
 *       </Dropdown>
 *       <Title>{selectedOption.name}</Title>
 *       <Text>{selectedOption.calories}</Text>
 *     </Provider>
 *   );
 * };
 *
 * export default MyComponent;
 * ```
 */
class Dropdown<T> extends React.Component<Props<T>, State> {
  static displayName = 'Dropdown';

  // @component ./DropdownOption.tsx
  static Option = DropdownOption;

  static getDerivedStateFromProps<T>(nextProps: Props<T>) {
    if (nextProps.selectedKey !== undefined) {
      return { selectedOption: nextProps.selectedKey };
    }
    return null;
  }

  popupViewRef: React.RefObject<View>;

  constructor(props: Props<T>) {
    super(props);

    this.popupViewRef = React.createRef();
    this.state = {
      isOpen: false,
      selectedKey: props.selectedKey ?? null,
      coordinates: { top: 0, left: 0, width: 0 },
    };
  }

  private setMenuOpen = (open: boolean) => {
    if (Platform.OS === 'web') {
      this.popupViewRef?.current?.measure(
        (_x, _y, width, height, pageX, pageY) =>
          this.setState({
            coordinates: {
              width: width,
              left: pageX,
              top: pageY + height,
            },
          })
      );
    }
    this.setState({ isOpen: open });
  };

  public openMenu = () => this.setMenuOpen(true);

  public closeMenu = () => this.setMenuOpen(false);

  public toggleMenuOpen = () => this.setMenuOpen(!this.state.isOpen);

  public options = () =>
    React.Children.map<OptionProps<T>, React.ReactElement<OptionProps<T>>>(
      this.props.children ?? [],
      (child) => child.props
    );

  public selectedOption = () =>
    this.options().find(
      (option) => option.optionKey === this.state.selectedKey
    ) ?? null;

  private renderOptions = (options: OptionProps<T>[]) => {
    const {
      emptyDropdownMessage = DEFAULT_EMPTY_DROPDOWN_LABEL,
      unselectLabel = this.props.placeholder ?? DEFAULT_PLACEHOLDER_LABEL,
      unselectIcon,
      required = false,
      onSelect,
    } = this.props;

    let optionsList = options.map((option) => (
      <DropdownChild
        key={option.optionKey}
        props={option}
        onPress={() => {
          this.setState({ selectedKey: option.optionKey });
          this.closeMenu();
          if (onSelect) {
            onSelect(option.value);
          }
        }}
      />
    ));

    if (!required) {
      optionsList.unshift(
        <DropdownChild
          key={DROPDOWN_NULL_OPTION_KEY}
          props={{
            value: undefined,
            optionKey: DROPDOWN_NULL_OPTION_KEY,
            label: unselectLabel,
            title: unselectLabel,
            left: unselectIcon,
            style: styles.unselectOption,
          }}
          onPress={() => {
            this.setState({ selectedKey: null });
            this.closeMenu();
            if (onSelect) {
              onSelect(null);
            }
          }}
        />
      );
    } else if (options.length === 0) {
      optionsList.push(
        <List.Item
          key={DROPDOWN_NULL_OPTION_KEY}
          disabled
          title={emptyDropdownMessage}
        />
      );
    }
    return optionsList;
  };

  private optionLabel = (option: OptionProps<T> | null) => {
    if (option) {
      return option.label ?? option.title;
    }
    return this.props.placeholder ?? DEFAULT_PLACEHOLDER_LABEL;
  };

  render() {
    const { maxHeight = DEFAULT_MAX_HEIGHT, theme } = this.props;
    const selectedOption = this.selectedOption();
    const options = this.options();

    let optionsChildren = this.renderOptions(options);

    return (
      <View ref={this.popupViewRef}>
        <TouchableRipple borderless onPress={this.toggleMenuOpen}>
          <Surface
            style={[
              styles.container,
              {
                borderRadius: theme.roundness,
                borderColor: this.state.isOpen
                  ? theme.colors.primary
                  : theme.colors.backdrop,
              },
            ]}
          >
            <Text
              style={[
                styles.text,
                {
                  color: selectedOption
                    ? theme.colors.text
                    : theme.colors.placeholder,
                },
              ]}
            >
              {this.optionLabel(selectedOption)}
            </Text>
            <IconButton
              style={styles.icon}
              icon="menu-down"
              onPress={this.toggleMenuOpen}
            />
          </Surface>
        </TouchableRipple>
        <Portal>
          {Platform.select({
            web: this.state.isOpen && (
              <TouchableWithoutFeedback
                style={styles.modalContainer}
                onPress={this.closeMenu}
              >
                <View style={styles.modal}>
                  <Surface
                    style={[this.state.coordinates, styles.modalContent]}
                  >
                    <ScrollView style={{ maxHeight: maxHeight }}>
                      {optionsChildren}
                    </ScrollView>
                  </Surface>
                </View>
              </TouchableWithoutFeedback>
            ),
            default: (
              <Dialog visible={this.state.isOpen} onDismiss={this.closeMenu}>
                <ScrollView>{optionsChildren}</ScrollView>
              </Dialog>
            ),
          })}
        </Portal>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    borderWidth: 1,
    alignItems: 'center',
    elevation: 1,
  },
  text: {
    flex: 1,
    paddingHorizontal: 10,
  },
  icon: {
    margin: 2.5,
  },
  modalContainer: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
  },
  modal: {
    width: '100%',
    height: '100%',
  },
  modalContent: {
    position: 'relative',
  },
  unselectOption: {
    opacity: 0.5,
  },
});

export default withTheme(Dropdown);
