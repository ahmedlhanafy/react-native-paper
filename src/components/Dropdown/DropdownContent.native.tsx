import React, { useContext } from 'react';
import { ScrollView } from 'react-native';
import { DropdownContext } from './Dropdown';
import Dialog from '../Dialog/Dialog';

type Props = {
  children: React.ReactNodeArray;
};

const DropdownContent = ({ children }: Props) => {
  const { closeMenu } = useContext(DropdownContext);

  return (
    <Dialog visible onDismiss={closeMenu}>
      <ScrollView>{children}</ScrollView>
    </Dialog>
  );
};

export default DropdownContent;
