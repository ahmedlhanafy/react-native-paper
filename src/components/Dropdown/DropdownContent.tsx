import Surface from '../Surface';
import {
  ScrollView,
  StyleSheet,
  TouchableWithoutFeedback,
  View,
} from 'react-native';
import React, { useContext } from 'react';
import { DropdownContext } from './Dropdown';

type Props = {
  children: React.ReactNode;
};

const DEFAULT_MAX_HEIGHT = 350;

const DropdownContent = (props: Props) => {
  const {
    closeMenu,
    dropdownCoordinates,
    maxHeight = DEFAULT_MAX_HEIGHT,
  } = useContext(DropdownContext);

  return (
    <TouchableWithoutFeedback onPress={closeMenu}>
      <View style={[StyleSheet.absoluteFill]}>
        <Surface style={[dropdownCoordinates, { maxHeight }]}>
          <ScrollView>{props.children}</ScrollView>
        </Surface>
      </View>
    </TouchableWithoutFeedback>
  );
};

export default DropdownContent;
