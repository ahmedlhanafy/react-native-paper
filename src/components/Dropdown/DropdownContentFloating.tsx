import Surface from '../Surface';
import {
  ScrollView,
  StyleSheet,
  TouchableWithoutFeedback,
  View,
} from 'react-native';
import React from 'react';
import { DropdownContext } from './Dropdown';
import { withTheme } from '../../core/theming';

type Props = {
  children: React.ReactNode;
  visible: boolean;
  theme: ReactNativePaper.Theme;
};

const DropdownContentFloating = ({ children, visible, theme }: Props) => {
  const { closeMenu, dropdownCoordinates } = React.useContext(DropdownContext);

  if (!visible) return null;

  return (
    <TouchableWithoutFeedback onPress={closeMenu}>
      <View style={[StyleSheet.absoluteFill]}>
        <Surface
          style={[
            dropdownCoordinates,
            styles.container,
            {
              borderBottomRightRadius: theme.roundness,
              borderBottomLeftRadius: theme.roundness,
            },
          ]}
        >
          <ScrollView>{children}</ScrollView>
        </Surface>
      </View>
    </TouchableWithoutFeedback>
  );
};

const styles = StyleSheet.create({
  container: {
    elevation: 8,
  },
});

export default withTheme(DropdownContentFloating);
