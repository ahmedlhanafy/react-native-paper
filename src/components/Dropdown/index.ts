import Dropdown from './Dropdown';
import DropdownOption from './DropdownOption';

export default Object.assign(
  // @component ./Dropdown.tsx
  Dropdown,
  {
    // @component ./DropdownOption.tsx
    Option: DropdownOption,
  }
);
