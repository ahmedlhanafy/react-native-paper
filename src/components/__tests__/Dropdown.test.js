import * as React from 'react';
import renderer from 'react-test-renderer';
import Dropdown from '../Dropdown/Dropdown';

it('renders an empty dropdown', () => {
  const tree = renderer
    .create(<Dropdown placeholder="Nothing here!" />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders dropdown with multiple options', () => {
  const tree = renderer.create(
    <Dropdown>
      <Dropdown.Option optionKey={1} value={1} title="Option 1" />
      <Dropdown.Option optionKey={2} value={2} title="Option 2" />
    </Dropdown>
  );

  expect(tree).toMatchSnapshot();
});

it('throws an error when Dropdown.Option is used outside dropdown', () => {
  const render = () => {
    renderer.create(
      <Dropdown.Option optionKey={1} value={1} title="Option1" />
    );
  };

  expect(render).toThrowErrorMatchingSnapshot();
});
