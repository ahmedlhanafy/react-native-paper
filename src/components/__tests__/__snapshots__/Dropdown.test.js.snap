// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an empty dropdown 1`] = `
<View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#ffffff",
          "borderColor": "rgba(0, 0, 0, 0.5)",
          "borderRadius": 4,
          "borderWidth": 1,
          "elevation": 1,
          "flexDirection": "row",
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 0.5,
            "width": 0,
          },
          "shadowOpacity": 0.24,
          "shadowRadius": 0.75,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
            },
            Array [
              Object {
                "flex": 1,
                "paddingHorizontal": 10,
              },
              Object {
                "color": "rgba(0, 0, 0, 0.54)",
              },
            ],
          ]
        }
      >
        Nothing here!
      </Text>
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "disabled": undefined,
          }
        }
        accessible={true}
        focusable={true}
        hitSlop={
          Object {
            "bottom": 6,
            "left": 6,
            "right": 6,
            "top": 6,
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Array [
              Object {
                "alignItems": "center",
                "justifyContent": "center",
                "margin": 6,
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 18,
                "height": 36,
                "width": 36,
              },
              undefined,
              Object {
                "margin": 2.5,
              },
            ],
          ]
        }
      >
        <View>
          <Text
            accessibilityElementsHidden={true}
            allowFontScaling={false}
            importantForAccessibility="no-hide-descendants"
            pointerEvents="none"
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontSize": 24,
                },
                Array [
                  Object {
                    "lineHeight": 24,
                    "transform": Array [
                      Object {
                        "scaleX": 1,
                      },
                    ],
                  },
                  Object {
                    "backgroundColor": "transparent",
                  },
                ],
                Object {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders dropdown with multiple options 1`] = `
<View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#ffffff",
          "borderColor": "rgba(0, 0, 0, 0.5)",
          "borderRadius": 4,
          "borderWidth": 1,
          "elevation": 1,
          "flexDirection": "row",
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 0.5,
            "width": 0,
          },
          "shadowOpacity": 0.24,
          "shadowRadius": 0.75,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
            },
            Array [
              Object {
                "flex": 1,
                "paddingHorizontal": 10,
              },
              Object {
                "color": "rgba(0, 0, 0, 0.54)",
              },
            ],
          ]
        }
      >
        Select an option
      </Text>
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "disabled": undefined,
          }
        }
        accessible={true}
        focusable={true}
        hitSlop={
          Object {
            "bottom": 6,
            "left": 6,
            "right": 6,
            "top": 6,
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Array [
              Object {
                "alignItems": "center",
                "justifyContent": "center",
                "margin": 6,
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 18,
                "height": 36,
                "width": 36,
              },
              undefined,
              Object {
                "margin": 2.5,
              },
            ],
          ]
        }
      >
        <View>
          <Text
            accessibilityElementsHidden={true}
            allowFontScaling={false}
            importantForAccessibility="no-hide-descendants"
            pointerEvents="none"
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontSize": 24,
                },
                Array [
                  Object {
                    "lineHeight": 24,
                    "transform": Array [
                      Object {
                        "scaleX": 1,
                      },
                    ],
                  },
                  Object {
                    "backgroundColor": "transparent",
                  },
                ],
                Object {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`throws an error when Dropdown.Option is used outside dropdown 1`] = `"<DropdownOption/> must be a direct child to <Dropdown/>"`;
